File: leetcode/python/0111_minimum_depth_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root):
        self.maximum = 1e8
        if root == None:
            return 0

        def helper(node, depth):
            if not node.left and not node.right:
                self.maximum = min(self.maximum, depth + 1)
            if node.left:
                helper(node.left, depth + 1)
            if node.right:
                helper(node.right, depth + 1)

        helper(root, 0)
        return self.maximum

File: leetcode/java/0111_minimum_depth_of_binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int helper(TreeNode node, int currDepth, int depth) {
        if (node != null) {
            currDepth++;
            depth = helper(node.left, currDepth, depth);

            if (currDepth < depth && (node.left == null && node.right == null)) depth = currDepth;

            depth = helper(node.right, currDepth, depth);
            currDepth--;
        }

        return depth;
    }

    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        return helper(root, 0, Integer.MAX_VALUE);
    }
}